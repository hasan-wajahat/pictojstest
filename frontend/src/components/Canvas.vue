<template>
  <div class="canvas col-sm-8 col-md-8 col-lg-8">
    <div class="block canvas" @dragover="onDragOver" @drop="onDrop">
      <template v-for="(asset, index) in assets">
        <div
          draggable="true"
          @dragstart="onDragStart(asset, $event)"
          @drop="onItemDrop(asset, $event)"
          @dragover="onDragOver"
          v-if="asset.type === 'image'"
          :key="asset.id"
          class="image"
          :style="getStyles(index)"
        >
          <img :src="asset.url">
          <button @click="$emit('delete-asset', index)" class="delete-button btn-danger">
            <i class="fa fa-close"></i>
          </button>
        </div>
      </template>
    </div>
  </div>
</template>

<script>
export default {
  name: "Canvas",
  props: {
    assets: Array
  },
  data: function() {
    return {
      movedObjectSize: {
        x: 0,
        y: 0
      }
    };
  },
  methods: {
    onDragStart: function(asset, event) {
      event.dataTransfer.setData("text", asset.id);
      this.movedObjectSize = {
        x: event.target.clientWidth,
        y: event.target.clientHeight
      };
    },
    onDragOver: function(event) {
      event.preventDefault();
    },
    onDrop: function(event) {
      event.preventDefault();
      const id = Number(event.dataTransfer.getData("text"));
      let offsetValues = {
        left: event.offsetX - this.movedObjectSize.x / 2,
        top: event.offsetY - this.movedObjectSize.y / 2
      };
      this.$emit("asset-moved", { id, offsetValues });
    },
    onItemDrop: function(asset, event) {
      event.preventDefault();
      event.stopPropagation();
      const id = Number(event.dataTransfer.getData("text"));
      let offsetValues = {
        left: asset.left + event.offsetX - this.movedObjectSize.x / 2,
        top: asset.top + event.offsetY - this.movedObjectSize.y / 2
      };
      this.$emit("asset-moved", { id, offsetValues });
    },
    getStyles: function(index) {
      return {
        left: `${this.assets[index].left}px`,
        top: `${this.assets[index].top}px`
      };
    }
  }
};
</script>

<style lang="scss" scoped>
.canvas {
  position: relative;
  overflow: hidden;
}
.image {
  cursor: pointer;
  position: absolute;
  img {
    max-width: 90%;
  }
  &:hover {
    background: darken(white, 5%);
    .delete-button {
      display: block;
    }
  }
}
.delete-button {
  padding: 2px 4px;
  position: absolute;
  top: -5px;
  right: -5px;
  font-size: 12px;
  border-radius: 50%;
  cursor: pointer;
  display: none;
}
</style>