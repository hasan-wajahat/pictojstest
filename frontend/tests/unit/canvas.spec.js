import { shallowMount } from '@vue/test-utils'
import Canvas from '@/components/Canvas.vue'

const ASSETS = [
  {
    "url": "http://localhost:8000/images/uploads-1462948498227.png",
    "type": "image",
    "id": 1560493937920,
    "left": 184,
    "top": 174
  },
  {
    "type": "text",
    "text": "testing text",
    "id": 1560493940607,
    "left": 228,
    "top": 204
  }
];

describe('Canvas.vue', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallowMount(Canvas, {
      propsData: {
        assets: ASSETS
      }
    });
  })

  it('renders container', () => {
    expect(wrapper.find('div').exists()).toBeTruthy();
  })

  it('renders correct number of images and text', () => {
    expect(wrapper.findAll('img').length).toBe(1);
    expect(wrapper.findAll('textarea').length).toBe(1);
  })

  it('can delete asset', () => {
    wrapper.find('.delete-button').trigger('click');
    expect(wrapper.emitted()).toMatchSnapshot();
  })

  it('can save assets', () => {
    wrapper.find('.save-button').trigger('click');
    expect(wrapper.emitted()).toMatchSnapshot();
  })
})
