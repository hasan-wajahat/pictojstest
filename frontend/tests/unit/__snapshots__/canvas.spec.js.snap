// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Canvas.vue can delete asset 1`] = `
Object {
  "delete-asset": Array [
    Array [
      0,
    ],
  ],
}
`;

exports[`Canvas.vue can save assets 1`] = `
Object {
  "save": Array [
    Array [],
  ],
}
`;
